### PR TITLE
GFORMS-1445 - Upload file to upscan ie. integration with upscan-initiate

### PR DESCRIPTION
**Pull request by josef-vlach**

### CODE REVIEW

File: app/uk/gov/hmrc/gform/upscan/UpscanController.scala [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+          )
+        upscanCallback <-
+          upscanCallbackPayload.asEither.leftMap(error =>
+            logger.error(s"Upscan callback payload failed to parse: ${request.body}, reported error: $error")
+          )
+
+      } yield upscanCallback match {
+        case upscanCallbackSuccess: UpscanCallback.Success =>
+          for {
+            file <- upscanService.download(upscanCallbackSuccess.downloadUrl)
```
Comment (girishkamat) [inline app/uk/gov/hmrc/gform/upscan/UpscanController.scala:93]: For large files, downloading as ByteString wont cause OOMs? Have you explored streaming if supported?

File: app/uk/gov/hmrc/gform/upscan/UpscanController.scala [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+            _    <- fileUploadFrontendAlgebra.uploadFile(envelopeId, fileId, upscanCallbackSuccess.uploadDetails, file)
+            form <- gformBackEndAlgebra.getForm(formIdData)
+            formUpd = setTransferred(form, formComponentId, fileId)
+            _                  <- gformBackEndAlgebra.updateUserData(formUpd, formIdData.maybeAccessCode)
+            upscanConfirmation <- upscanService.confirm(upscanCallbackSuccess)
```
Comment (girishkamat) [inline app/uk/gov/hmrc/gform/upscan/UpscanController.scala:98]: is the formatting for for compression a bit messed up?

File: app/uk/gov/hmrc/gform/upscan/UpscanController.scala [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+          } yield NoContent
```
Comment (girishkamat) [inline app/uk/gov/hmrc/gform/upscan/UpscanController.scala:99]: After updating the user data in forms collection, if upscan confirm fails, we need to rollback the update isnt it?

File: app/uk/gov/hmrc/gform/upscan/UpscanService.scala [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+  upscanConnector: UpscanConnector,
+  upscanRepository: UpscanRepository,
+  gformConnector: GformConnector,
+  queryParameterCrypto: CryptoWithKeysFromConfig,
+  appConfig: AppConfig
+)(implicit ec: ExecutionContext)
+    extends UpscanAlgebra[Future] {
+
+  def upscanInitiate(
+    fileUploadIds: List[FormComponentId],
```
Comment (girishkamat) [inline app/uk/gov/hmrc/gform/upscan/UpscanService.scala:43]: why is this a collection, I was under the assumption its always going to be a single file per page?

File: app/uk/gov/hmrc/gform/upscan/UpscanService.scala [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+      successRedirect,
+      errorRedirect
+    )
+  }
+
+  def download(
+    downloadUrl: String
+  ): Future[ByteString] = upscanConnector.download(downloadUrl)
+
+  def waitForConfirmation(reference: UpscanReference): Future[UpscanFileStatus] =
```
Comment (girishkamat) [inline app/uk/gov/hmrc/gform/upscan/UpscanService.scala:109]: Why call this waitForConfirmation? Is the find method going to be blocked until a value appears in the mongo collection?

Comment (josef-vlach) [reply app/uk/gov/hmrc/gform/upscan/UpscanService.scala]: renamed to `retrieveConfirmationOrFail`

**Review by girishkamat — APPROVED** [review]

**Pull request merged**